master/internal/core_checkpoint.go

```diff
@@ -0,0 +1,397 @@
+	aw archiveWriter
+	// Internal states
+	err    error
+	pos    int
+	bucket string
+}
+
+// Next() returns true if the next item is available.
+func (i *batchDownloadIterator) Next() bool {
+	i.pos++
+	if i.pos == len(i.objects) {
+		return false
+	}
+	err := i.aw.WriteFileHeader(*i.objects[i.pos].Key, *i.objects[i.pos].Size)
+	if err != nil {
+		i.err = err
+		return false
+	}
+	return true
+}
+
+// Err() eturns the error if any.
+func (i *batchDownloadIterator) Err() error {
+	return i.err
+}
+
+// DownloadObject() eturns a DownloadObject.
+func (i *batchDownloadIterator) DownloadObject() s3manager.BatchDownloadObject {
+	return s3manager.BatchDownloadObject{
+		Object: ptrs.Ptr(s3.GetObjectInput{
+			Bucket: ptrs.Ptr(i.bucket),
+			Key:    i.objects[i.pos].Key,
+		}),
+		Writer: newS3SeqWriterAt(i.aw),
+	}
+}
+
+func newBatchDownloadIterator(aw archiveWriter,
+	bucket string, objs []*s3.Object) *batchDownloadIterator {
+	return &batchDownloadIterator{
+		aw:      aw,
+		bucket:  bucket,
+		objects: objs,
+		pos:     -1,
+	}
+}
+
+func getAwsRegion() string {
+	defaultRegion := "us-west-2"
+	client := http.Client{
+		Timeout: 100 * time.Millisecond,
+	}
+
+	// Per AWS: https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instance-identity-documents.html
+	resp, err := client.Get("http://169.254.169.254/latest/dynamic/instance-identity/document")
+	if err != nil || resp.StatusCode != http.StatusOK {
+		return defaultRegion
+	}
+
+	bytes, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return defaultRegion
+	}
+	_ = resp.Body.Close()
+
+	var jsonObj map[string]interface{}
+	err = json.Unmarshal(bytes, &jsonObj)
+	if err != nil {
+		return defaultRegion
+	}
+
+	if _, ok := jsonObj["region"]; !ok {
+		return defaultRegion
+	}
+
+	region, ok := jsonObj["region"].(string)
+	if !ok {
+		return defaultRegion
+	}
+
+	return region
+}
+
+func s3DownloadCheckpoint(
+	c echo.Context, aw archiveWriter, bucket string, prefix string,
+) error {
+	sess, err := session.NewSession(&aws.Config{
+		Region: aws.String(getAwsRegion()),
+	})
+	if err != nil {
+		return err
+	}
+	// We do not pass in credentials explicitly. Instead, we reply on
+	// the existing AWS credentials.
+	s3client := s3.New(sess)
+
+	var errs []error
+	downloader := s3manager.NewDownloader(sess, func(d *s3manager.Downloader) {
+		d.Concurrency = 1 // Setting concurrency to 1 to use s3SeqWriterAt
+	})
+	funcReadPage := func(output *s3.ListObjectsV2Output, lastPage bool) bool {
+		iter := newBatchDownloadIterator(aw, bucket, output.Contents)
+		// Download every bucket in this page
+		err = downloader.DownloadWithIterator(c.Request().Context(), iter)
+		if iter.Err() != nil {
+			errs = append(errs, iter.Err())
+		}
+		if err != nil {
+			errs = append(errs, err)
+		}
+
+		// Return False to stop paging
+		return len(errs) == 0
+	}
+	err = s3client.ListObjectsV2PagesWithContext(
+		c.Request().Context(),
+		&s3.ListObjectsV2Input{
+			Bucket: ptrs.Ptr(bucket),
+			Prefix: ptrs.Ptr(prefix),
```

Comment:
Just use `&bucket` and `&prefix`, `ptrs.Ptr` is for cases where the go language makes it difficult to get a pointer, like `ptrs.Ptr(true)` because `&true` is not allowed.

When you already have a `string` on the stack like `bucket`, `ptrs.Ptr(bucket)` is going to create a copy of it on the heap and pass a pointer to that.

Reply:
Makes sense. Don't know what I was thinking...